### PR TITLE
Port versioned namespace logic from org.wildfly:wildfly-clustering-common

### PR DESCRIPTION
**Pull request by pferraro**
See https://issues.redhat.com/browse/WFLY-12775

### CODE REVIEW

File: src/main/java/org/jboss/staxmapper/XMLMapper.java [hunk 1]
```diff
@@ -124,5 +124,20 @@ private Factory() {
...
         }
+
+        /**
+         * Convenience method that Creates a new mapper instance pre-registered with a set of schemas, each supplying an element reader.
+         * @param <T> the operating type of the reader
+         * @param <S> the schema type
+         * @param schemas the schemas with which the returned mapper will be auto-registered
+         * @return a new mapper instance, pre-configured with the specified schemas
+         */
+        public static <T, S extends Schema & Supplier<XMLElementReader<T>>> XMLMapper create(Iterable<S> schemas) {
```
Comment (dmlloyd) [inline src/main/java/org/jboss/staxmapper/XMLMapper.java:135]: Is a `Schema` always expected to also be a `Supplier<XMLElementReader<?>>`? This seems at odds with the above JavaDoc example, where a `Schema` was more expected to map to an `XMLElementReader` by way of a secondary function.

I would at least consider either changing the `XMLElementReader` to be more schema-aware, or make `Schema` more of a data object.

Comment (pferraro) [reply src/main/java/org/jboss/staxmapper/XMLMapper.java]: @dmlloyd  I reworked this a bit based on your feedback.  Specifically:
* Split-up Schema into 2 interfaces:
  1. Namespace - constructs a namespace URI.  Added type parameter to narrow usage of Namespace.since(...) to classes of the same type.
  1. QNameProvider - creates a QName from a Namespace and a local name
* Added a separate XMLElementReaderFactory interface that supplies a reader for a specific element QName
* Replaced previous XMLMapper.Factory.create(...) with a XMLMapper.registerRootElement(...) convenience method

**Comment by pferraro** [comment]
@jmesnil Please review.

File: src/main/java/org/jboss/staxmapper/Namespace.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+
+package org.jboss.staxmapper;
+
+/**
+ * Defines a versioned XML namespace.
+ * @author Paul Ferraro
+ */
+public interface Namespace<N extends Namespace<N>> {
```
Comment (dmlloyd) [inline src/main/java/org/jboss/staxmapper/Namespace.java:29]: This would signify an object which is-a namespace, okay so far, I guess this is an object representation of the information contained in the namespace URI?

File: src/main/java/org/jboss/staxmapper/QNameProvider.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+package org.jboss.staxmapper;
+
+import javax.xml.XMLConstants;
+import javax.xml.namespace.QName;
+
+/**
+ * Provides the qualified name of an XML element or attribute.
+ * @author Paul Ferraro
+ */
+public interface QNameProvider<N extends Namespace<N>> {
```
Comment (dmlloyd) [inline src/main/java/org/jboss/staxmapper/QNameProvider.java:32]: This is a little bit confusing, in part because it seems reasonable on the surface, but it's not clear how a qname is being provided here, or why... does this represent a function which takes a namespace and emits a qname for it? Where does it get the namespace context to know what qname prefix(es) go with what namespaces?

**Review by dmlloyd — COMMENTED** [review]
Honestly I still find this API a little bit confusing in its complexity. By analyzing the code I guess the goals are:

* Have an object type which encapsulates our notion of a URN, typically a name and version pair
* Have a version comparison mechanism for the URN representation
* Allow registration of a reader based on that URN representation rather than a QName
* Provide a single interface to read the XML structure which also knows what URN (actually QName) it parses

Did I miss any goals?

File: src/main/java/org/jboss/staxmapper/Namespace.java (hunk 2 above)
Comment (pferraro) [reply src/main/java/org/jboss/staxmapper/Namespace.java:29]: The purpose is to encapsulate the info comprising the namespace URI, and provide a means to compare namespaces for use by an XMLElementReader.
The general expectation is that user would be implement this interface via an enum, which would represent the set of namespaces supported by a given reader.

File: src/main/java/org/jboss/staxmapper/QNameProvider.java (hunk 3 above)
Comment (pferraro) [reply src/main/java/org/jboss/staxmapper/QNameProvider.java:32]: This is meant to encapsulate the components of a QName. The wiring of this association is up to the implementor.  In the simplest case, the Namespace implementation (e.g. an enum) would also implement QNameProvider, and thus additionally associate the namespace with a common local name.

**Comment by pferraro** [comment]
> Honestly I still find this API a little bit confusing in its complexity. By analyzing the code I guess the goals are:
> 
> * Have an object type which encapsulates our notion of a URN, typically a name and version pair
> * Have a version comparison mechanism for the URN representation
> * Allow registration of a reader based on that URN representation rather than a QName
> * Provide a single interface to read the XML structure which also knows what URN (actually QName) it parses
>
> Did I miss any goals?

Points 1, 2, and 4 are goals.  The 3rd point, as you've summarized, is not really a goal, but rather incorporates the proposed interfaces into the existing API.

**Comment by pferraro** [comment]
@dmlloyd  I've revised this proposal again based on your previous feedback.  Let me know what you think.
I've also generalized the "version" associated with a versioned namespace, and added separate common implementations.

File: src/main/java/org/jboss/staxmapper/NaryVersion.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    int[] values() {
+        return this.values;
+    }
+
+    @Override
+    public int compareTo(T version) {
+        int[] values = version.values();
+        assert this.values.length == values.length;
```
Comment (dmlloyd) [inline src/main/java/org/jboss/staxmapper/NaryVersion.java:48]: This assertion enforces an invariant that I don't see anywhere on the class. It might be better to have a rule like "shorter versions sort earlier" or "missing trailing segments are zeros" or something like that.

**Review by dmlloyd — APPROVED** [review]
By and large, I like this a lot better; it's a lot clearer what the purpose of each thing is, and each thing is better-encapsulated.

I think it could still possibly be improved slightly by making the `NaryVersion` class just be a final class e.g. just `Version` or `NumericalVersion`; it avoids a little messiness around potentially misbehaving subclasses (experience has taught me that if you give subclassers an inch, they will surely take a mile).  Then you could replace the public constructor with a factory method which gives you more future flexibility without costing the user.  By adding the suggested accessors, and maybe also even adding `getMajor` etc. to the superclass which default to 0 if no corresponding segment is present, it becomes quite a bit simpler to handle.

That said, I don't want to hold this up any longer over bikeshed-type commentary (you've been more than patient) so consider this further review under "advisement" and take or leave what you want. 🙂

**Comment by pferraro** [comment]
> I think it could still possibly be improved slightly by making the NaryVersion class just be a final class e.g. just Version or NumericalVersion; it avoids a little messiness around potentially misbehaving subclasses (experience has taught me that if you give subclassers an inch, they will surely take a mile). Then you could replace the public constructor with a factory method which gives you more future flexibility without costing the user. By adding the suggested accessors, and maybe also even adding getMajor etc. to the superclass which default to 0 if no corresponding segment is present, it becomes quite a bit simpler to handle.

Initially, NaryVersion was a package protected base implementation, where Binary/TernaryVersion were the public (and ideally final) implementations - but in the end, I opted to make it public given that the class itself is non-abstract.  The general idea was that you would only compare a BinaryVersion with another BinaryVersion (hence the variable type of NaryVersion).  However, having read your feedback and thinking on this some more, this restriction is not necessary, e.g. 1.0 and 1.0.0 are semantically equivalent.

In summary, I like your suggestion about scrapping the subclasses for a sufficiently generic (and better named) concrete version implementation.  I'll push an update later today.

File: src/main/java/org/jboss/staxmapper/NaryVersion.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+ * A version based on N terms.
+ * @author Paul Ferraro
+ */
+public class NaryVersion<T extends NaryVersion<T>> implements Comparable<T> {
+
+    private final int[] values;
+
+    public NaryVersion(int... values) {
+        this.values = values;
+    }
```
[the remark this answers is not in the document]
Comment (pferraro) [reply src/main/java/org/jboss/staxmapper/NaryVersion.java]: Good point.  Fixed.

File: src/main/java/org/jboss/staxmapper/NaryVersion.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(this.values);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public boolean equals(Object object) {
+        return (this == object) || (this.getClass().isInstance(object) ? this.compareTo((T) object) == 0 : false);
```
[the remark this answers is not in the document]
Comment (pferraro) [reply src/main/java/org/jboss/staxmapper/NaryVersion.java]: Feedback incorporated.

File: src/main/java/org/jboss/staxmapper/NaryVersion.java (hunk 4 above)
Comment (pferraro) [reply src/main/java/org/jboss/staxmapper/NaryVersion.java]: Feedback incorporated.

File: src/main/java/org/jboss/staxmapper/NaryVersion.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    @Override
+    public String toString() {
+        return IntStream.of(this.values).mapToObj(String::valueOf).collect(Collectors.joining("."));
```
[the remark this answers is not in the document]
Comment (pferraro) [reply src/main/java/org/jboss/staxmapper/NaryVersion.java]: I've added these methods.

File: src/main/java/org/jboss/staxmapper/NaryVersion.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    public NaryVersion(int... values) {
+        this.values = values;
+    }
+
+    int[] values() {
+        return this.values;
+    }
+
+    @Override
+    public int compareTo(T version) {
```
[the remark this answers is not in the document]
Comment (pferraro) [reply src/main/java/org/jboss/staxmapper/NaryVersion.java]: I've made this class final.

File: src/main/java/org/jboss/staxmapper/NaryVersion.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+import java.util.Objects;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * A version based on N terms.
+ * @author Paul Ferraro
+ */
+public class NaryVersion<T extends NaryVersion<T>> implements Comparable<T> {
+
```
[the remark this answers is not in the document]
Comment (pferraro) [reply src/main/java/org/jboss/staxmapper/NaryVersion.java]: I've removed the self-type - which was meant to be defined by subtypes, e.g. BinaryVersion vs TernaryVersion; which are now scrapped.

I've added a method to override the comparator of a Versioned object for use by the since(...) method.  This enables the use of smallrye-common-version via something like:

```java
enum Version implements Versioned<String, MyNamespace> {
    V1_0("1.0"),
    V1_1("1.1"),
    ;
    private final String version;

    Version(String version) {
        this.version = version;
    }

    @Override
    public String getVersion() {
        return this.version;
    }

    @Override
    public Comparator<String> getComparator() {
        return VersionSchema.BASIC;
    }
}
```

While this works, I've kept IntVersion, which peforms faster comparisons.

**Comment by pferraro** [comment]
@dmlloyd Thanks again for the feedback.
NaryVersion is now IntVersion, is no longer type parameterized, and no longer allows subclassing.
A Versioned object can also override the Comparator used by Versioned.since(...) when comparing versions, to support Versioned<String, V> via smallrye-commons-version.
Let me know if you have any other feedback.

File: src/main/java/org/jboss/staxmapper/IntVersion.java [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+        return 0;
+    }
+
+    @Override
+    public int hashCode() {
+        // Ignore trailing zero segments
+        int length = this.segments.length;
+        while ((length > 0) && (this.segments[length -1] == 0)) {
+            length -= 1;
+        }
```
Comment (dmlloyd) [inline src/main/java/org/jboss/staxmapper/IntVersion.java:151]: It might be simpler to trim trailing zeros in the constructor. You could do this in the array constructor using `Arrays.copyOf(original, cnt)` to trim off as many zeros as you find.

Comment (pferraro) [reply src/main/java/org/jboss/staxmapper/IntVersion.java]: Indeed.  It also renders the fixed arg constructors redundant.
Fixed.

**Comment by pferraro** [comment]
@dmlloyd Assuming this is ready to be merged now, who should contact re: merging and releasing?

**Comment by dmlloyd** [comment]
Well it looks like the last person to release this was... me, six years ago... so I guess... me? 😅
